pyro/nn/module.py

```diff
@@ -83,29 +108,37 @@ def forward(self):
         dims and no subsampling will be performed.
     """
 
+    init_value: Optional[Union[torch.Tensor, Callable[[], torch.Tensor]]] = None
+    constraint: constraints.Constraint = constraints.real
+    event_dim: Optional[int] = None
+
     # Support use as a decorator.
-    def __get__(self, obj, obj_type):
+    def __get__(
+        self, obj: Optional["PyroModule"], obj_type: Type["PyroModule"]
+    ) -> "PyroParam":
         assert issubclass(obj_type, PyroModule)
         if obj is None:
             return self
 
-        name = self.init_value.__name__
+        name = self.init_value.__name__  # type: ignore[union-attr]
         if name not in obj.__dict__["_pyro_params"]:
             init_value, constraint, event_dim = self
-            init_value = functools.partial(init_value, obj)  # bind method's self arg
+            # bind method's self arg
+            init_value = functools.partial(init_value, obj)  # type: ignore[arg-type]
             setattr(obj, name, PyroParam(init_value, constraint, event_dim))
-        return obj.__getattr__(name)
+        value: PyroParam = obj.__getattr__(name)
+        return value
 
     # Support decoration with optional kwargs, e.g. @PyroParam(event_dim=0).
-    def __call__(self, init_value):
+    def __call__(
+        self, init_value: Union[torch.Tensor, Callable[[], torch.Tensor]]
+    ) -> "PyroParam":
         assert self.init_value is None
         return PyroParam(init_value, self.constraint, self.event_dim)
 
 
-PyroParam.__new__.__defaults__ = (None, constraints.real, None)
-
-
-class PyroSample(namedtuple("PyroSample", ("prior",))):
+@dataclass
```

Comment:
Can we use `@dataclass(slots=True, frozen=True)` here, to remain closer to namedtuple? Of course we'll then need to use `object._setattribute__(self, "foo", bar)` rather than `self.foo = bar` in `__post_init__`.

Reply:
`slots` seems to be new in Python 3.10